src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -87,10 +84,11 @@ internal static IntPtr AllocateSslContext(long options, SafeX509Handle certHandl
             }
             catch
             {
-                Marshal.StructureToPtr(sslContext, sslContextPtr, false);
-                FreeSslContext(sslContextPtr);
-                return IntPtr.Zero;
+                Disconnect(sslContext.sslPtr);
+                throw;
             }
+
+            IntPtr sslContextPtr = Marshal.AllocHGlobal(Marshal.SizeOf<SslContext>());
```

Comment:
yes, that will be done separately. This PR address some of the comments, not all.